chiselFrontend/src/main/scala/chisel3/core/Binding.scala

```diff
@@ -2,203 +2,89 @@ package chisel3.core
-      target,
-      element => element.binding match {
-        case unbound @ UnboundBinding(_) => {}
-        case binding => throw AlreadyBoundException(binding.toString)
-      }
-    )
-    catch {
-      case BindingException(message) => throw BindingException(s"$error_prelude$message")
+/** Requires that a node is a chisel type (not hardware, "unbound")
+  */
+object requireIsChiselType {
+  def apply(node: Data, msg: String = "") = if (node.hasBinding) {
+    if (!msg.isEmpty()) {
+      throw new Binding.BindingException(s"$msg '$node' must be a Chisel type, not hardware")
+    } else {
+      throw new Binding.BindingException(s"'$node' must be a Chisel type, not hardware")
     }
   }
+}
 
-  // Excepts if any root element is unbound and thus not on the hardware graph
-  def checkSynthesizable(target: Data, error_prelude: String): Unit = {
-    try walkToBinding(
-      target,
-      element => {
-        // Compatibility mode to automatically wrap ports in IO
-        // TODO: remove me, perhaps by removing Bindings checks from compatibility mode
-        element._parent match {
-          case Some(x: BaseModule) => x._autoWrapPorts
-          case _ =>
-        }
-        // Actual binding check
-        element.binding match {
-          case SynthesizableBinding() => // OK
-          case binding => {
-            // Attempt to diagnose common bindings issues, like forgot to wrap IO(...)
-            element._parent match {
-              case Some(x: LegacyModule) =>
-                // null check in case we try to access io before it is defined
-                if ((x.io != null) && (x.io.flatten contains element)) {
-                  throw MissingIOWrapperException
-                }
-              case _ =>
-            }
-            // Fallback generic exception
-            throw NotSynthesizableException
-          }
-        }
+// Element only direction used for the Binding system only.
+sealed abstract class BindingDirection
+object BindingDirection {
+  case object Internal extends BindingDirection  // internal type or wire
+  case object Output extends BindingDirection  // module port
+  case object Input extends BindingDirection  // module port
+
+  def from(binding: TopBinding, direction: ActualDirection) = {
+    binding match {
+      case PortBinding(_) => direction match {
+        case ActualDirection.Output => Output
+        case ActualDirection.Input => Input
+        case dir => throw new RuntimeException(s"Unexpected port element direction '$dir'")
       }
-    )
-    catch {
-      case BindingException(message) => throw BindingException(s"$error_prelude$message")
+      case _ => Internal
     }
   }
 }
 
 // Location refers to 'where' in the Module hierarchy this lives
 sealed trait Binding {
   def location: Option[BaseModule]
-  def direction: Option[Direction]
 }
+// Top-level binding representing hardware, not a pointer to another binding
+sealed trait TopBinding extends Binding
```

Comment:
By "top" here do you mean just a bound type inside of a module? Children are fields in an aggregate? Top kind of misleads me here since we usually refer to "Top" as the top of the circuit but I'm not sure what a better word would be....